launcher/container_runner.go

```diff
@@ -194,43 +208,112 @@ func (r *ContainerRunner) measureContainerClaims(ctx context.Context) error {
 			return err
 		}
 	}
 	return nil
 }
 
-// Run the container
-// Doesn't support container restart yet
-// Container output will always be redirected to stdio for now
-func (r *ContainerRunner) Run(ctx context.Context) error {
-	if err := r.measureContainerClaims(ctx); err != nil {
-		return fmt.Errorf("failed to measure container claims: %v", err)
+func getTTL(token []byte) (time.Duration, error) {
+	claims := &jwt.RegisteredClaims{}
+	_, _, err := new(jwt.Parser).ParseUnverified(string(token), claims)
+	if err != nil {
+		return 0, fmt.Errorf("failed to parse token: %w", err)
+	}
+
+	now := time.Now()
+	if !now.Before(claims.ExpiresAt.Time) {
+		return 0, errors.New("token is expired")
+	}
+
+	return time.Until(claims.ExpiresAt.Time), nil
+}
+
+func (r *ContainerRunner) refreshToken(ctx context.Context, refreshMultiplier float64) (time.Duration, error) {
+	if refreshMultiplier > 1 {
+		return 0, fmt.Errorf("refresh multiplier cannot exceed 1, got %v", refreshMultiplier)
+	} else if refreshMultiplier <= 0 {
+		return 0, fmt.Errorf("refresh multiplier must be positive, got %v", refreshMultiplier)
 	}
 
-	// Create hostTokenPath directory. Should come before NewTask.
 	token, err := r.attestAgent.Attest(ctx)
 	if err != nil {
-		return fmt.Errorf("failed to retrieve attestation service token: %v", err)
+		return 0, fmt.Errorf("failed to retrieve attestation service token: %v", err)
+	}
+
+	filepath := path.Join(HostTokenPath, attestationVerifierTokenFile)
+	if err = os.WriteFile(filepath, token, 0644); err != nil {
+		return 0, fmt.Errorf("failed to write token to container mount source point: %v", err)
+	}
+
+	ttl, err := getTTL(token)
+	if err != nil {
+		return 0, fmt.Errorf("failed to get token TTL: %v", err)
 	}
 
-	// Write OIDC token for the container.
-	if err := os.MkdirAll(HostTokenPath, 0644); err != nil {
+	return time.Duration(float64(ttl) * refreshMultiplier), nil
+}
+
+func (r *ContainerRunner) fetchAndWriteToken(ctx context.Context) error {
+	if err := os.MkdirAll(HostTokenPath, 0744); err != nil {
 		log.Fatal(err)
 	}
-	if err = os.WriteFile(path.Join(HostTokenPath, attestationVerifierTokenFile), token, 0644); err != nil {
-		return fmt.Errorf("failed to write token to container mount source point: %v", err)
+
+	duration, err := r.refreshToken(ctx, defaultRefreshMultiplier)
+	if err != nil {
+		return err
+	}
+
+	// If timer is already configured.
+	if r.tokenRefresher.timer != nil {
+		r.tokenRefresher.timer.Stop()
+	}
+
+	// Set a timer to refresh the token before it expires.
+	r.tokenRefresher.timer = time.NewTimer(duration)
+	r.tokenRefresher.done = make(chan bool)
+	go func() {
+		for {
+			select {
+			case <-r.tokenRefresher.done:
+				return
+			case <-r.tokenRefresher.timer.C:
+				// Refresh token.
+				duration, err := r.refreshToken(ctx, defaultRefreshMultiplier)
+				if err != nil {
+					log.Printf("failed to refresh attestation service token: %v", err)
+					continue
```

Comment:
If we aren't going to reset the timer, should this be a `return` to make the behavior more clear?